### ISSUE TITLE
Unable to execute command

### ISSUE THREAD

**Issue by half-nothing** [UnioDex/VelocityRcon#6]
Hi.I have found a problem when I use [mcrcon](https://github.com/Pigeon-Server/mcrcon-withcpp) to connect velocity.
I have tried on the vanilla server and I'm pretty sure it is effective.
But when I use command "glist all" and "lpv help".The command line is not responding.I take a screen shot and the pictures are below.

The Vanilla Server:
![`7 }ZQHN)Y6}HHOX8DM4I$N](https://user-images.githubusercontent.com/50048293/167255220-97fd8e83-3351-45da-abae-dd23eddfbff3.png)

Velocity:
![@1$K1DRHXZH{}VF(T`3}H3W](https://user-images.githubusercontent.com/50048293/167255231-4d69b845-3d3a-4b50-a510-a8ea2d4d6228.png)
![T(Q}PG`@WF)KKF3DO5U71AL](https://user-images.githubusercontent.com/50048293/167255234-4d4dc52a-8fbb-4eea-be5b-bd484de92930.png)

So,I use wireshark to capture and analyze packets.And I find this.
![2I0YM(`BVP`R`WAN4XUC)MG](https://user-images.githubusercontent.com/50048293/167255283-97702302-0564-45cc-8d6a-b383aea3e12f.png)
It sent back my instructions verbatim!
So,what happened?

**Comment by louboecker** [comment]
I'm having the same issue, it won't let me use any commands

Edit: Apparently it does execute commands, but I just don't get the response

**Comment by isaaclepes** [comment]
When using mrcon, Velocity does not respond to 'end' command. (or any command, really).
I tried waiting, but after about 15 minutes it was obviously not going to shut down.
When using mrcon, after entering a command, you are not returned to a ">" prompt for issuing another command, it simply hangs on a new line with a blinking cursor.
Further, when connected via mrcon and I simply attempt the 'quit' command to disconnect, the session is not ended by VelocityRcon.
When using mrcon from command line to send a command, the mrcon process remains alive until Velocity is stopped.

mrcon functions fine on my other 4 papermc servers as well as my old Waterfall instance using BungeeRcon.

latest.log only shows the following for attempts:
[nioEventLoopGroup-5-1/INFO]: Rcon connection from [/[0:0:0:0:0:0:0:1]:55260]

**Comment by Lucifer2033** [comment]
hi I have the same problem it executes the command but no response is coming

**Comment by IntangibleW** [comment]
The same to me.
I got no response from Rcon and log shows there is Rcon port listening:
![屏幕截图_20230226_100053](https://user-images.githubusercontent.com/43871076/221388211-58357d4c-28d5-47fe-8c14-55a871ee1202.png)
![屏幕截图_20230226_100131](https://user-images.githubusercontent.com/43871076/221388213-42aa08f1-0557-4c06-80ed-6479a45e8b5d.png)
I use velocity-3.2.0-SNAPSHOT-230

**Comment by GuilhermeWerner** [comment]
I have the same problem here. I run velocity as a daemon on a linux server and I need a way to send commands to it to reload the config. RCON would be perfect but this problem won't let me.

**Comment by MeesJ** [comment]
Same issue here

**Comment by bitter-geek** [comment]
So... is this plugin abandoned? I'm having the same issue mentioned above when trying to connect from a variety of different RCON clients. With no communication from the developer since this bug was initially reported over a year ago, should I start looking around for another option?

**Comment by GuilhermeWerner** [comment]
I forked it to try to resolve this bug, but in the end I just added Gradle support and for me it worked.

If anyone wants to test: https://github.com/TribuFu-Forks/VelocityRcon

**Comment by MeesJ** [comment]
> I forked it to try to resolve this bug, but in the end I just added Gradle support and for me it worked.
> 
> If anyone wants to test: https://github.com/TribuFu-Forks/VelocityRcon

Cool! Would it be possible for you to compile it as a jar? I'll test it tomorrow.

**Comment by bitter-geek** [comment]
I agree, a jar file would be awesome for folks like me that don't speak Java. Also, what is Gradle?

**Comment by GuilhermeWerner** [comment]
> > I forked it to try to resolve this bug, but in the end I just added Gradle support and for me it worked.
> > If anyone wants to test: https://github.com/TribuFu-Forks/VelocityRcon
> 
> Cool! Would it be possible for you to compile it as a jar? I'll test it tomorrow.

I created a test release: https://github.com/TribuFu-Forks/VelocityRcon/releases/tag/v.1.2-TEST

> I agree, a jar file would be awesome for folks like me that don't speak Java. Also, what is Gradle?

It's a build system for Java, I prefer it to Maven, it's no big deal.

**Comment by bitter-geek** [comment]
Oh, gotcha. Thank you! I'll play around with it soon, too. I've wasted two days trying to figure this out only to find this bug report today. It's nice knowing where the problem is now, but frustrating that it's been known for this long without a fix.

**Comment by bitter-geek** [comment]
@GuilhermeWerner I tried to compile this code with some instructions I found online but I'm not having much luck. Again, I don't speak Java. Can you explain how to compile this or maybe just throw a jar file into the repo? Thanks for your help!

**Comment by GuilhermeWerner** [comment]
> @GuilhermeWerner I tried to compile this code with some instructions I found online but I'm not having much luck. Again, I don't speak Java. Can you explain how to compile this or maybe just throw a jar file into the repo? Thanks for your help!

I left a .jar in the link I sent: https://github.com/TribuFu-Forks/VelocityRcon/releases/tag/v.1.2-TEST

![image](https://github.com/ygtdmn/VelocityRcon/assets/26710260/9602d613-c411-45b1-a95b-393ba61c577a)

**Comment by bitter-geek** [comment]
Wow. I don't know how I missed that... Sorry to bug you and thanks again!

**Comment by half-nothing** [comment]
> > @GuilhermeWerner I tried to compile this code with some instructions I found online but I'm not having much luck. Again, I don't speak Java. Can you explain how to compile this or maybe just throw a jar file into the repo? Thanks for your help!
> 
> I left a .jar in the link I sent: https://github.com/TribuFu-Forks/VelocityRcon/releases/tag/v.1.2-TEST
> 
> ![image](https://user-images.githubusercontent.com/26710260/279453007-9602d613-c411-45b1-a95b-393ba61c577a.png)

You solved the problem?

**Comment by GuilhermeWerner** [comment]
> You solved the problem?

Technically I didn't change anything in the code. I forked it to investigate, debugged it and used gradle to compile.

After that I tested it on my servers and it worked, I've been using this "modified" version for months, and for me it worked, but I only recompiled the project using gradle, nothing else.

That's why I asked if more people could test it.

**Comment by half-nothing** [comment]
> > You solved the problem?
> 
> Technically I didn't change anything in the code. I forked it to investigate, debugged it and used gradle to compile.
> 
> After that I tested it on my servers and it worked, I've been using this "modified" version for months, and for me it worked, but I only recompiled the project using gradle, nothing else.
> 
> That's why I asked if more people could test it.

Maybe you should see this
https://github.com/PaperMC/Velocity/issues/833
Anyway, I'll try the Vanilla Server's commands.

**Comment by half-nothing** [comment]
> > You solved the problem?
> 
> Technically I didn't change anything in the code. I forked it to investigate, debugged it and used gradle to compile.
> 
> After that I tested it on my servers and it worked, I've been using this "modified" version for months, and for me it worked, but I only recompiled the project using gradle, nothing else.
> 
> That's why I asked if more people could test it.

![image](https://github.com/ygtdmn/VelocityRcon/assets/50048293/d5855954-5b52-4b11-8678-d28e37828714)
The problem is still there, maybe I can look into it some time

**Comment by bitter-geek** [comment]
I'm able to send some commands with the Gradle version, but not others. The `velocity version` command works:
![image](https://github.com/ygtdmn/VelocityRcon/assets/145600012/f782eb3c-7263-455c-8f55-67a3912bf057)

But the `glist` command returns this:
![image](https://github.com/ygtdmn/VelocityRcon/assets/145600012/fbc51f59-83ac-44c8-b747-97b1b2f75339)

... when it should return this:
![image](https://github.com/ygtdmn/VelocityRcon/assets/145600012/c3fbebc7-ebdd-48af-8f32-e49c20248c21)

So we're getting closer! I'm also not able to use any commands for any plugins, it says "no such command" when the same command works fine in the Velocity console.

**Comment by aws1313** [comment]
For me it is fixed here: https://github.com/TISUnion/VelocityRcon

**Comment by MeesJ** [comment]
> For me it is fixed here: https://github.com/TISUnion/VelocityRcon

All issues? Including the issue @bitter-geek reported?

**Comment by aws1313** [comment]
I think so... It works with [mcrcon](https://github.com/Tiiffi/mcrcon) and also with [mcipc](https://github.com/conqp/mcipc) for me.
`velocity version` :

![image](https://github.com/ygtdmn/VelocityRcon/assets/69417093/faeef2d9-9508-4854-94aa-d18eae844bd8)
`glist`:

![image](https://github.com/ygtdmn/VelocityRcon/assets/69417093/c3e67d3c-4c40-4c25-8474-240232f6722d)
`glist all`:

![image](https://github.com/ygtdmn/VelocityRcon/assets/69417093/5095b2ec-f8f5-4be8-9a74-eeb6d322c20d)


You can try it out yourself. I used https://github.com/TISUnion/VelocityRcon/releases/download/v1.1.1/VelocityRcon.jar

The issue with `lpv` still persists though. I have to send `lpv help`, wait a little bit an then send any other command in order to get the result of `lpv help` and if I understood it right this can't be fixed:
![image](https://github.com/ygtdmn/VelocityRcon/assets/69417093/dfcca0d0-dafb-498f-af40-f20c1714fc92)

**Comment by aws1313** [comment]
Using [mcipc](https://github.com/conqp/mcipc) you could work around this with something like:
```python
from mcipc.rcon.je import Client
import time


def run_command(cmd, *args):
    with Client("HOST", PORT, passwd="PASSWORD") as c:
        c.run(cmd, *args)
        time.sleep(0.2)
        return '\n'.join(c.run("glist",).splitlines()[:-2])


if __name__ == '__main__':
    res = run_command("lpv", "help")
    print(res)

``` 
```
[LP] Running LuckPerms v5.4.102.
> /lpv user <user>
> /lpv group <group>
> /lpv track <track>
> /lpv log
> /lpv sync
> /lpv info
> /lpv editor [type]
> /lpv verbose <on|record|off|upload> [filter]
> /lpv tree [scope] [player]
> /lpv search <permission>
> /lpv networksync
> /lpv import <file>
> /lpv export <file>
> /lpv reloadconfig
> /lpv bulkupdate
> /lpv translations
> /lpv creategroup <group>
> /lpv deletegroup <group>
> /lpv listgroups
> /lpv createtrack <track>
> /lpv deletetrack <track>
> /lpv listtracks
```

**Comment by DefomCode** [comment]
Is this going to be decided? I tried all the solutions that I found here, but nothing helped me, there is no connection to rcon at all.

**Comment by bitter-geek** [comment]
@DefomCode I don't think this original version is ever going to be fixed. My guess is the developer has abandoned the project.

@aws1313 I'll try this soon. Is this your fork? I'm curious what was done to make things work.

**Comment by aws1313** [comment]
> @DefomCode I don't think this original version is ever going to be fixed. My guess is the developer has abandoned the project.
> 
> @aws1313 I'll try this soon. Is this your fork? I'm curious what was done to make things work.

@bitter-geek No, it is not my fork, but I have looked at the code. It seems that problem was that Velocity's responses work with ComponentMessages from  [KyoriPowered/adventure](https://github.com/KyoriPowered/adventure), but this plugin only uses raw strings.
So the things your `glist` command returned like `velocity.command.glist-player-plural` or `velocity.command.glist-view-all` were some sort of ComponentMessage identifier/key. The plugin originally just returned them but they should have been interpreted/used by the ComponentApi. In the fixed version (which I linked) the identifier/key is taken, formatted to the correct message and then returned.

**Comment by bitter-geek** [comment]
@aws1313 Interesting! I gave it a try and the core commands work, but I'm not having any luck with plugin commands. I don't think I'm doing anything wrong, but I'm still pretty new to all of this.

When I run the plugin command from the console itself it works fine, but when I run it from rcon it says: `Error executing: rc (TextComponentImpl{content="No such command", style=StyleImpl{obfuscated=not_set, bold=not_set, strikethrough=not_set, underlined=not_set, italic=not_set, color=NamedTextColor{name="red", value="#ff5555"}, clickEvent=null, hoverEvent=null, insertion=null, font=null}, children=[]})`

**Comment by aws1313** [comment]
@bitter-geek Can you give a little bit more information about which commands you are trying and which rcon client you are using? It still works fine for me...

**Comment by ygtdmn** [comment]
Hey guys, I stopped Minecraft development years ago and this repo is not active anymore. However, I can accept any pull requests and also give full control to a new maintainer. Hit me up on twitter/YigitDuman

**Comment by bitter-geek** [comment]
@aws1313 Sure, I'm trying to use [RustyConnector](https://github.com/Aelysium-Group/rusty-connector) to dynamically add/remove servers to the proxy (necessary for my environment but not working well with Waterfall). I'm also using [mcrcon](https://github.com/Tiiffi/mcrcon) for the commands. 

It could be some kind of wonky permissions thing (I've seen strange cases where RCON can't run commands in server plugins) but I don't know how to resolve it if that's the case.

**Comment by aws1313** [comment]
@bitter-geek I was able to reproduce your problem. However, in this case the error is not caused by VelocityCron, but by RustyConnector itself.
The way RustyConnector processes commands results in only commands from the console being accepted and executed. In addition, the responses are always sent to the console.
I have created [a fork of RustyConnector](https://github.com/aws1313/rusty-connector) and fixed the error there. If you use the [.jar file](https://github.com/aws1313/rusty-connector/releases/download/v0.7.1/rusty-connector-0.7.1.jar) from the release ["RustyConnector v0.7.1 for VelocityRcron"](https://github.com/aws1313/rusty-connector/releases/tag/v0.7.1) it should work. The main problem was
```java
.requires(source -> source instanceof ConsoleCommandSource)
``` 
in line 40 in ["/plugin/velocity/central/command/CommandRusty.java"](https://github.com/Aelysium-Group/rusty-connector/blob/34a623ec2911e1cc8e87fb6461ec48e755a35f51/plugin/velocity/src/main/java/group/aelysium/rustyconnector/plugin/velocity/central/command/CommandRusty.java) in the [original repository](https://github.com/Aelysium-Group/rusty-connector) which forced the command to be sent from console.

**Comment by bitter-geek** [comment]
@aws1313 Dang! You're a wizard! I'll give this a try soon. I really appreciate your help!

**Comment by bitter-geek** [comment]
Hey @aws1313, I gave the RustyConnector .jar you made a try but it says I can't run it because it was compiled in a version of Java that's too high for my system. I'm running version 17. I know you're not really supporting this plugin, but if you have a few spare minutes and could re-compile the changes you made in a version that's <= 17 that would be awesome.

**Comment by aws1313** [comment]
@bitter-geek  Oops. I'm sorry... I recompiled it with Java 17 and created a [new release](https://github.com/aws1313/rusty-connector/releases/tag/v0.7.2) and you can download the correct .jar file [here](https://github.com/aws1313/rusty-connector/releases/download/v0.7.2/rusty-connector-0.7.2.jar). I hope it works this time...

**Comment by bitter-geek** [comment]
@aws1313 no need to be sorry! I just appreciate the help. It appears to be working now! Thank you! Do you have like a... "buy me a coffee" link or anything?

**Comment by aws1313** [comment]
@bitter-geek No, I don't have a "Buy me a coffee" or anything like that, but I don't need that anyways. I'm just glad I could help you (and maybe others who will read this in the future) :)

**Comment by MarkapToGo** [comment]
Is there now a working version that also handles responses from luckperms?
